woocommerce/woocommerce
# Add slotFill for variation menus

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR adds slotFills for the `Quick update` menus to enhance extensibility. It also refactors them to remove some code duplication we had.

### Extensibility

**_Quick update_ actions**
```
<VariationQuickUpdateMenuItem
    group= { 'pricing' }
    supportsMultipleSelection={ true }
    order={ 20 }
>
    // Dropdown, MenuGroup, MenuItem
</VariationQuickUpdateMenuItem>
```

When `supportsMultipleSelection` is `true`, it will extend the menu shown after pressing `Quick actions`, which means that menu items will show up in both the single selection and multiple selection menus. Otherwise, it will only extend the single update menu.

Closes #42737.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure the feature `New product editor` is enabled under `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features` and the feature `product-variation-management` is enabled under `Features` tab from `/wp-admin/tools.php?page=woocommerce-admin-test-helper`.
3. Go to Products > Add New > Variations (`/wp-admin/admin.php?page=wc-admin&path=/add-product&tab=variations`) and add some attributes to generate variations.
4. Wait until variations get generated
5. Verify the actions under the vertical ellipsis are working well.
6. Select at least one variation and verify the Quick update actions are working well too.
7. Now it's time to test extensibility. You can checkout the branch `dev/42737_slotfill_for_variation_menus_test`. I added a few fields to different parts of the variation actions menus. [This extension](https://github.com/woocommerce/woocommerce/files/13973436/myplugin.zip) will add the following components


## Update

You can see the components [here](https://gist.github.com/octaedro/db58dd4377e1f57157080fd624925a28).

It adds:
- an item with the text `** Top level item 2 **` at the top level of the single update. It has order 20, so it should be rendered after the other top-level item.
- an item with the text `** Top level item **` at the top level of the single and multiple updates with order 10.
- an item with the text `** Pricing item / Set price **` into the pricing group for single and multiple updates. After pressing it, it changes the regular price for a random number. It works for single and multiple variations.
- an item with the text `** Shipping item **` into the shipping group of single updates. It doesn't trigger the onClose callback.
- an item with the text `** Downloads item / Set download limit **` into the downloads group of single updates. After pressing it, it updates the download item with a random number.
- an item with the text `** Inventory item / Set stock status: outofstock **` into the inventory group of single and multiple updates. After pressing it, it modifies the stock status to `out of stock`. It works for single and multiple variations.

Every added item will print in the browser's console:
```
[click] -> Plugin top level
Type - [ "single" or "multiple and single" ] items 
Selected items: [ selection ]
[ a quick description ]
```


### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
